projects/ngx-sub-form/src/lib/ngx-root-form.component.ts

```diff
@@ -0,0 +1,113 @@
+import { isNil } from 'lodash-es';
+import { NgxSubFormRemapComponent } from './ngx-sub-form.component';
+import { takeUntilDestroyed } from './ngx-sub-form-utils';
+
+export abstract class NgxRootFormComponent<ControlInterface, FormInterface = ControlInterface>
+  extends NgxSubFormRemapComponent<ControlInterface, FormInterface>
+  implements OnInit, OnChanges {
+  public abstract dataInput: Required<ControlInterface> | null;
+  public abstract dataOutput: EventEmitter<ControlInterface | null>;
+  // using a private variable `_dataOutput$` to be able to control the
+  // emission rate with a debounce for ex
+  /** @internal */
+  protected _dataOutput$: Subject<ControlInterface | null> = new Subject();
+
+  protected dataInput$: Subject<ControlInterface | null> = new Subject();
+  protected updateTimeoutMs = 500;
+
+  protected emitInitialValueOnInit = false;
+  protected emitNullOnDestroy = false;
+
+  protected dataValue: ControlInterface | null = null;
+
+  /** @internal */
+  protected applyingChangeSignal = new BehaviorSubject(false);
+
+  protected applyingChange$ = this.applyingChangeSignal.pipe(
+    switchMap(isApplying => {
+      if (!isApplying) {
+        return of(false);
+      }
+
+      return timer(this.updateTimeoutMs).pipe(
+        mapTo(false),
+        startWith(true),
+      );
+    }),
+    distinctUntilChanged(),
+  );
+
+  public ngOnInit(): void {
+    // we need to manually call registerOnChange because that function
+    // handles most of the logic from NgxSubForm and when it's called
+    // as a ControlValueAccessor that function is called by Angular itself
+    this.registerOnChange(data => {
+      if (this.formGroup.invalid || isEqual(data, this.dataInput)) {
+        return;
+      }
+
+      this.dataValue = data;
+    });
+
+    this.dataInput$
+      .pipe(
+        tap(() => this.applyingChangeSignal.next(false)),
+        filter(newValue => !isNil(newValue) && !isEqual(newValue, this.formGroup.value)),
+        tap(newValue => this.writeValue(newValue as Required<ControlInterface>)),
+        takeUntilDestroyed(this),
+      )
+      .subscribe();
+
+    this.applyingChange$
+      .pipe(
+        tap(isApplying => this.handleApplyingChangeTimeout(isApplying)),
+        takeUntilDestroyed(this),
+      )
+      .subscribe();
+
+    this._dataOutput$
+      .pipe(
+        // hook to give access to the observable for sub-classes
+        this.handleDataOutput(),
+        tap(value => this.dataOutput.emit(value)),
+        takeUntilDestroyed(this),
+      )
+      .subscribe();
+  }
+
+  // from NgxRootFormComponent we simply return the observable and let sub-classes override that method if needed
+  protected handleDataOutput(): (obs$: Observable<ControlInterface | null>) => Observable<ControlInterface | null> {
+    return obs$ => obs$;
+  }
+
+  public ngOnChanges(): void {
+    this.dataInput$.next(this.dataInput);
```

Comment:
This should probably be inspecting the changes value for the actual dataInput key changing as the extending component could have a bunch of inputs unrelated to the form

Reply:
Good catch! Fixed